### ISSUE TITLE
Should use_vignette() also add rmarkdown in VignetteBuilder?

### ISSUE THREAD

**Issue by maelle** [r-lib/usethis#1147]
As [brought to my attention](https://github.com/r-hub/blog/pull/113#discussion_r433193417) by @gaborcsardi, Writing R Extensions [states](https://cran.r-project.org/doc/manuals/R-exts.html#The-DESCRIPTION-file) _"Note that if, for example, a vignette has engine ‘knitr::rmarkdown’, then knitr provides the engine but both knitr and rmarkdown are needed for using it, so both these packages need to be in the ‘VignetteBuilder’ field and at least suggested (as rmarkdown is only suggested by knitr, and hence not available automatically along with it). "_

**Comment by jennybc** [comment]
I (and I think many in tidyverse / r-lib) use `use_vignette()` heavily and I don't feel like I hear of actual suffering from *not* putting rmarkdown in that field.

Is this issue motivated by some actual incident where it's been a problem? What is typical across CRAN packages?

**Comment by maelle** [comment]
No actual problem, hence the question mark. 

`pkgsearch::advanced_search(VignetteBuilder = "knitr AND rmarkdown")` showed several thousands packages. (AFK hence the absence of reprex)

**Comment by jennybc** [comment]
> pkgsearch::advanced_search(VignetteBuilder = "knitr AND rmarkdown") showed several thousands packages.

How do you see this? `advanced_search()` doesn't seem to be a function in pkgsearch and `available.packages()` doesn't seem to actually provide info on VignetteBuilder.

``` r
library(tidyverse)
#> Warning: package 'tibble' was built under R version 3.6.2
av <- available.packages(fields = c("VignetteBuilder", "Encoding"))
av <- as_tibble(av)
av %>% 
  select(Package, Version, VignetteBuilder)
#> # A tibble: 15,684 x 3
#>    Package     Version VignetteBuilder
#>    <chr>       <chr>   <chr>          
#>  1 A3          1.0.0   <NA>           
#>  2 aaSEA       1.1.0   <NA>           
#>  3 ABACUS      1.0.0   <NA>           
#>  4 abbyyR      0.5.5   <NA>           
#>  5 abc         2.1     <NA>           
#>  6 abc.data    1.0     <NA>           
#>  7 ABC.RAP     0.9.0   <NA>           
#>  8 abcADM      1.0     <NA>           
#>  9 ABCanalysis 1.2.1   <NA>           
#> 10 abcdeFBA    0.4     <NA>           
#> # … with 15,674 more rows
av %>% 
  count(VignetteBuilder)
#> # A tibble: 1 x 2
#>   VignetteBuilder     n
#> * <chr>           <int>
#> 1 <NA>            15684
```

<sup>Created on 2020-06-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by maelle** [comment]
Which pkgsearch version are you using?

**Comment by maelle** [comment]
Much less specific search https://github.com/search?l=&q=knitr%2C+rmarkdown+user%3Acran+filename%3ADESCRIPTION&type=Code

**Comment by maelle** [comment]
Mmmh so the `pkgsearch` call above _does_ return 3812 packages but when I look their DESCRIPTION files they don't have both packages, sorry. (the advanced search is too advanced for me :upside_down_face: )

Using [Julia's approach](https://juliasilge.com/blog/mining-cran-description/ ), there are only 46 packages with both packages under VignetteBuilder.

``` r
cran <- tools::CRAN_package_db()

# the returned data frame has two columns with the same name???
cran <- cran[,-65]

# make it a tibble
cran <- tibble::as_tibble(cran)

dplyr::count(cran, VignetteBuilder, sort = TRUE)
#> # A tibble: 29 x 2
#>    VignetteBuilder             n
#>    <chr>                   <int>
#>  1 <NA>                    11101
#>  2 knitr                    4279
#>  3 R.rsp                     198
#>  4 knitr, rmarkdown           46
#>  5 knitr, R.rsp               38
#>  6 utils, knitr               10
#>  7 R.rsp, knitr                9
#>  8 rmarkdown, knitr            9
#>  9 utils                       9
#> 10 knitr, rmarkdown, R.rsp     6
#> # … with 19 more rows
```

<sup>Created on 2020-06-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by maelle** [comment]
66 actually 

``` r
cran <- tools::CRAN_package_db()

# the returned data frame has two columns with the same name???
cran <- cran[,-65]

# make it a tibble
cran <- tibble::as_tibble(cran)

dplyr::count(cran, VignetteBuilder, sort = TRUE)
#> # A tibble: 29 x 2
#>    VignetteBuilder             n
#>    <chr>                   <int>
#>  1 <NA>                    11101
#>  2 knitr                    4279
#>  3 R.rsp                     198
#>  4 knitr, rmarkdown           46
#>  5 knitr, R.rsp               38
#>  6 utils, knitr               10
#>  7 R.rsp, knitr                9
#>  8 rmarkdown, knitr            9
#>  9 utils                       9
#> 10 knitr, rmarkdown, R.rsp     6
#> # … with 19 more rows
dplyr::filter(
  cran,
  grepl("rmarkdown", VignetteBuilder),
  grepl("knitr", VignetteBuilder)
)$Package
#>  [1] "AMR"                   "BIGL"                  "ciTools"              
#>  [4] "CoopGame"              "cyphr"                 "data.tree"            
#>  [7] "denoiSeq"              "drugCombo"             "eikosograms"          
#> [10] "ergm"                  "fitHeavyTail"          "gasfluxes"            
#> [13] "GFD"                   "ggTimeSeries"          "gprofiler2"           
#> [16] "grainscape"            "idiogramFISH"          "imputeFin"            
#> [19] "LTRCtrees"             "MANOVA.RM"             "MiRKAT"               
#> [22] "NetLogoR"              "nlpred"                "odr"                  
#> [25] "optband"               "PCMBase"               "PCMBaseCpp"           
#> [28] "PMCMRplus"             "polyCub"               "polyRAD"              
#> [31] "portfolioBacktest"     "POUMM"                 "ppclust"              
#> [34] "prettydoc"             "QFASA"                 "qualvar"              
#> [37] "quickPlot"             "radiomics"             "RcppNumerical"        
#> [40] "reproducible"          "riskParityPortfolio"   "robvis"               
#> [43] "RSpectra"              "SeqKat"                "sequenza"             
#> [46] "showtext"              "sim1000G"              "SpaDES"               
#> [49] "SpaDES.addins"         "SpaDES.core"           "sparseIndexTracking"  
#> [52] "spathial"              "spectralGraphTopology" "SSDM"                 
#> [55] "stylest"               "SubgrPlots"            "supc"                 
#> [58] "SuperLearner"          "sweidnumbr"            "tergm"                
#> [61] "utf8"                  "VCA"                   "VFP"                  
#> [64] "visNetwork"            "wrProteo"              "zooaRch"
```

<sup>Created on 2020-06-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by maelle** [comment]
Off-topic, this reminded me me of https://github.com/r-lib/remotes/issues/459 -- not that it'd help with CRAN and vignettes.

I'm still wondering what to recommend in my R-hub blog post regarding knitr and rmarkdown in `DESCRIPTION`, but these posts have less consequences than usethis helpers. :smile:

**Comment by jennybc** [comment]
Turns out I had a very old version of pkgsearch. I've upgraded now and have `advanced_search()`. I don't have a strong opinion but it seems like just including knitr is working fine for an awful lot of packages. So I'm inclined to stick with the status quo.

``` r
library(pkgsearch)
library(tidyverse)

x <- advanced_search("VignetteBuilder: knitr", size = 4800)
x %>% 
  # work around an infelicity with the `pkg_search_result` class
  as_tibble() %>% 
  hoist(package_data, "VignetteBuilder") %>% 
  mutate(
    knitr = grepl("knitr", VignetteBuilder),
    rmarkdown = grepl("rmarkdown", VignetteBuilder)
  ) %>% 
  count(knitr, rmarkdown) %>% 
  mutate(n / sum(n))
#> # A tibble: 2 x 4
#>   knitr rmarkdown     n `n/sum(n)`
#> * <lgl> <lgl>     <int>      <dbl>
#> 1 TRUE  FALSE      4653     0.985 
#> 2 TRUE  TRUE         72     0.0152
```

<sup>Created on 2020-06-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by maelle** [comment]
The only case where it'd be an issue would be if CRAN decided to check packages including vignettes, without installing `Suggests` dependencies?

In the blog post I'll remain vague regarding what usethis does ("handling dependencies aspect" or so).

**Comment by maelle** [comment]
Well it seems `remotes::install_deps(dependencies = TRUE)` won't install the `VignetteBuilder` packages so as you can imagine it doesn't go well on Travis &co. :sweat_smile:

**Comment by jennybc** [comment]
Thanks for doing so much analysis. I'm certainly more aware of the VignetteBuilder field now! For the moment, I don't see a clear need to change what usethis does and there's always some risk of unintended consequences when changing things that aren't broken. So I'm going to close.

**Comment by m-Py** [comment]
Hello all, 

I do not want re-open this issue or argue that the `VignetteBuilder` field should change to also include `rmarkdown` when using `use_vignette()`, I just want to report a behavior of `use_vignette()` that I noticed:

In a package of mine, I include both `rmarkdown` and `knitr` in the `VignetteBuilder` field of my DESCRIPTION file. Out of convenience, I used `use_vignette()` to create a new vignette, which then removed the `rmarkdown` package from the `VignetteBuilder` field. I do not think this is a big deal because I could just reverse it, I just wanted to highlight this behavior. Thanks for your work on the `usethis` package!

Cheers, Martin
